cmd/kubectl-datadog/agent/upgrade/upgrade.go

```diff
@@ -0,0 +1,174 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-2020 Datadog, Inc.
+
+package upgrade
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/DataDog/datadog-operator/pkg/apis/datadoghq/v1alpha1"
+	"github.com/DataDog/datadog-operator/pkg/plugin/common"
+
+	"github.com/spf13/cobra"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/cli-runtime/pkg/genericclioptions"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var (
+	image          string
+	latest         bool
+	latestImage    = "datadog/agent:latest"
+	upgradeExample = `
+  # upgrade the version of the datadog agent to latest
+  %[1]s upgrade --latest
+
+  # upgrade the version of the datadog agent defined in DatadogAgent named foo to latest
+  %[1]s upgrade foo --latest
+
+  # upgrade the datadog agent with a custom image
+  %[1]s upgrade --image <account>/<repo>:<tag>
+`
+)
+
+// options provides information required by agent upgrade command
+type options struct {
+	genericclioptions.IOStreams
+	configFlags          *genericclioptions.ConfigFlags
+	args                 []string
+	client               client.Client
+	userNamespace        string
+	userDatadogAgentName string
+}
```

Comment:
maybe you can create another struct `common.Options` 

```golang
package common

type Options struct {
        ConfigFlags          *genericclioptions.ConfigFlags
	Client               client.Client
	UserNamespace        string
	UserDatadogAgentName string
}

// SetNamespace configures the namespace
func (o *Options) SetNamespace(ns string) {
	o.userNamespace = ns
}

// SetClient configures the client if the field exists
func (o *Options) SetClient(client client.Client) {
	o.Client = client
}

// SetClientset configures the clientset if the field exists
func (o *Options) SetClientset(clientset *kubernetes.Clientset) {
}

// GetClientConfig returns the client config
func (o *Options) GetClientConfig() clientcmd.ClientConfig {
	return o.ConfigFlags.ToRawKubeConfigLoader()
}

/// another file
package upgrade

type options struct {
	genericclioptions.IOStreams
        common.Options
	args                 []string
}
```

Reply:
I actually thought about this but `SetClient` and `SetClientset` are not the same for all the commands, some commands have only one client type, and some commands have both